feat: free form code cells

boandriy opened the pull request:
Created a functionality to rearrange code cells on workflow, add cells, delete cells.

Commit:
feat: free form code cells

ramfox commented on src/features/workflow/WorkfolwHeading.tsx:
```diff
@@ -0,0 +1,23 @@
+import React from "react";
```
this filename has a spelling error:
"WorkfolwHeading" => "WorkflowHeading"

ramfox replied:
Actually, keeping in line with our other naming conventions, I think this name should be changed to `WorkflowHeader.tsx`

ramfox commented on src/features/workflow/WorkfolwHeading.tsx:
```diff
@@ -0,0 +1,23 @@
+import React from "react";
+
+
+interface WorkflowHeadingProps {
```
change to `WorkflowHeaderProps`

ramfox commented on src/features/workflow/WorkfolwHeading.tsx:
```diff
@@ -0,0 +1,23 @@
+  onChange: (t: string) => void;
+}
+
+const WorkflowHeading: React.FC<WorkflowHeadingProps> = ({
```
Change to `WorkflowHeader`

ramfox commented on src/features/workflow/WorkfolwHeading.tsx:
```diff
@@ -0,0 +1,23 @@
+     type="text"/>)
+}
+
+export default WorkflowHeading
```
change to `WorkflowHeader`

ramfox commented on src/features/workflow/WorkflowCell.tsx:
```diff
@@ -14,35 +16,45 @@ export interface WorkflowCellProps {
+//     default:
+//       return ''
+//   }
+// }
```
If we plan on bringing this back in the future, then add a TODO about when we will bring it back, otherwise, remove it

b5 replied:
remove commented-out code. we have git history for this

ramfox commented on src/features/workflow/WorkflowCell.tsx:
```diff
@@ -52,24 +64,44 @@ const WorkflowCell: React.FC<WorkflowCellProps> = ({
+        {/*    }}>{name}</h3>*/}
+        {/*    <div className='text-xs mb-2.5 text-gray-400'>{description}</div>*/}
+        {/*  </div>*/}
+        {/*</header>*/}
```
Either remove or add a TODO about when we are bringing this back.

ramfox commented on src/features/workflow/WorkflowScriptStatus.tsx:
```diff
@@ -0,0 +1,68 @@
+import React from "react";
+
+import {NewRunStep, Run, RunStatus} from "../../qri/run";
+import {Dataset, TransformStep} from "../../qri/dataset";
```
If you have a linter this should get fixed, but we have format our imports with spaces between the curly braces, and no ending `;`:

```
import React from "react"

import { NewRunStep, Run, RunStatus } from "../../qri/run"
import { Dataset, TransformStep } from "../../qri/dataset"
import WorkflowStepStatusCircle from "./WorkflowStepStatusCircle"
```
...

ahh, it looks like we don't have `ts-standard` installed.

filed an issue here https://github.com/qri-io/frontend/issues/339

ramfox commented on src/features/workflow/state/workflowActions.ts:
```diff
@@ -64,12 +68,20 @@ function fetchRunNow(qriRef: QriRef): ApiAction {
 }
 
-export interface SetWorkflowStepAction {
+export interface RemoveWorkflowStepAction {
   type: string
   index: number
+}
+
+
+export interface SetWorkflowStepAction extends RemoveWorkflowStepAction{
   script: string
 }
 
+export interface AddWorkflowStepAction extends RemoveWorkflowStepAction{
+  syntax: string
+}
+
```
For future readers, it may seem confusing why we are extending something called `RemoveWorkflowStepAction` for so many actions.

It's better to rename `RemoveWorkflowStepAction` to `WorkflowStepAction`, since it is being used for many more actions than just removing.

```
export interface WorkflowStepAction {
  type: string
  index: number
}

export interface SetWorkflowStepAction extends WorkflowStepAction{
  script: string
}

export interface AddWorkflowStepAction extends WorkflowStepAction{
  syntax: string
}
```

Then rename any interface returns from `RemoveWorkflowStepAction` to `WorkflowStepAction`

b5 commented on src/chrome/CustomIcon.tsx:
```diff
@@ -38,7 +40,8 @@ const CustomIcon: React.FunctionComponent<CustomIconProps> = ({
       width={dimension}
       height={dimension}
-      viewBox="0 0 24 24"
+      transform={`${rotation ? 'rotate('+rotation+')' : ''}`}
```
don't use the turnery operator to provide empty values, use `&&` instead:

```suggestion
      transform={rotation && `rotate(${rotation})`}
```

b5 commented on src/chrome/icon/Duplicate.tsx:
```diff
@@ -0,0 +1,14 @@
+export default Duplicate
+
+
+
```
aim to have at most one blank line terminate files

b5 commented on src/features/workflow/WorkflowCellControlButton.tsx:
```diff
@@ -0,0 +1,32 @@
+      onClick={onClick}>
+      <Icon
+        className='mr-2'
+        rotation={rotate ? 180 : undefined}
```
```suggestion
        rotation={rotate && 180}
```

b5 commented on src/features/workflow/WorkflowEditor.tsx:
```diff
@@ -7,11 +7,18 @@ import WorkflowTriggersEditor from '../trigger/WorkflowTriggersEditor'
+  duplicateWorkflowTransformStep,
+  clearOutputWorkflowTransformStep,
+  moveWorkflowTransformStepUp, moveWorkflowTransformStepDown
+} from './state/workflowActions'
```
smarter pattern here is to hook `WorkflowControls` into dispatch & import there. Props passed to `WorkflowControls` here should be the necessary arguments to dispatch actions, so the step index

b5 commented on src/features/workflow/WorkflowOutline.tsx:
```diff
@@ -53,50 +53,51 @@ const WorkflowOutline: React.FC<WorkflowOutlineProps> = ({
+          {/*      Click <span className='border rounded-sm px-1'>Dry Run</span> to preview the result of this script*/}
+          {/*    </div>*/}
+          {/*  )}*/}
+          {/*</ScrollTrigger>*/}
```
remove instead of commenting out

ramfox replied:
leave a TODO saying when we should add this back, or remove the comments entirely

b5 commented on src/features/workflow/WorkflowScriptStatus.tsx:
```diff
@@ -0,0 +1,68 @@
+          return (
+            <WorkflowStepStatusCircle status={r ? (r.status || 'waiting') : ''} />
+          )
+        })}
```
so the little dots don't correlate to the step execution (but that is really interesting!), instead they're supposed to indicate changes across components.

Instead I'd opt to hide these boxes for now

chriswhong committed:
refactor: update default workflow template code

b5 committed:
fix(workflow editor): remove commit preview dots

preview "scan tron" dots are intended to represent component changes across versions, hide them for now

chriswhong committed:
refactor: update default workflow template code

Merge pull request #340 from qri-io/cw/new-template-code
